### PR TITLE
feat(runner): fix '.only()' exclusive feature, #1481

### PR DESCRIPTION
**Pull request by a8m**
This PR replace #1591 plus I added some integration-tests for it(thanks @danielstjules) 
IMO we should merge it to **v3.0.0** asap, instead of rebase-it on master every time.

/cc @mochajs/mocha

[Review on Reviewable](https://reviewable.io/reviews/mochajs/mocha/1807)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -117,6 +117,22 @@ test-only:
 		--ui qunit \
 		test/acceptance/misc/only/qunit
 
+test-global-only:
```
Comment (boneskull) [inline Makefile]: @a8m these paths are incorrect

**Comment by boneskull** [comment]
@a8m after correcting that, please merge.  thanks!!

**Comment by dasilvacontin** [comment]
I just noticed that there's duplicate code in the interfaces that could be refactored into `Suite#setOnly(flag)` and `Test#setOnly(flag)` (or similar).

**Comment by a8m** [comment]
PTAL @boneskull
@dasilvacontin can you please be more specific ? thx

**Comment by dasilvacontin** [comment]
Welcome back @a8m! :smile:

```js
mocha.options.hasOnly = true;
suite.isOnly = true;
suite.onlyTests = (suite.onlyTests || []).concat(test);
```
and
```js
mocha.options.hasOnly = true;
suite.isOnly = true;
```
are repeated multiple times through the code.

I was proposing refactoring the first one into `Suite#addOnlyTest`. And the second one into `Suite#only` (similar to the setters/getters `Suite#timeout`, `Suite#slow`, `Suite#bail`, etc)

(and `Suite#addOnlyTest` would internally execute `Suite#only`, besides adding the test to the array, ofc)

**Comment by dallonf** [comment]
What's still missing in this PR? Is it something relatively simple that I could finish?

Also, is this meant to fix the issue where `describe.only` overrides any `it.only` implementations? (This is where I wound up when searching for that issue) Example:

```js
describe.only("These should run", function() {
  it("this one should not run (but does)", function() { });
  it.only("this one should run", function() { });
});

describe("These should not run", function() {
  it("and of course don't", function() {})
  it.only("although I wonder about this one...?", function() {})
});
```

**Comment by boneskull** [comment]
@a8m What is the status of this?  @danielstjules / @dasilvacontin Either of you guys know?

**Comment by dasilvacontin** [comment]
I believe it was lacking follow up of my comments from @a8m.

**Comment by a8m** [comment]
Hi guys! sorry for this huge delay :/
I made some changes according to @dasilvacontin's comments; would you mind to take a look at it?
/cc @boneskull  @dasilvacontin

**Comment by boneskull** [comment]
merged into v3.0.0 branch via 8a754344db3d4d85780572d38ada272b1faa6ee6

@a8m thanks!

**Comment by a8m** [comment]
Awesome!! thanks @boneskull

**Pull request closed without merging**